libdrgn/program.c

```diff
@@ -1489,6 +1522,86 @@ drgn_thread_object(struct drgn_thread *thread, const struct drgn_object **ret)
 	return NULL;
 }
 
+static struct drgn_error *
+drgn_thread_name_linux_kernel(struct drgn_thread *thread, char **ret)
+{
+	struct drgn_error *err;
+	DRGN_OBJECT(comm, drgn_object_program(&thread->object));
+	err = drgn_object_member_dereference(&comm, &thread->object, "comm");
+	if (!err)
+		err = drgn_object_read_c_string(&comm, ret);
+	if (err)
+		return err;
+	return NULL;
+}
+
+static struct drgn_error *
+drgn_thread_name_userspace_live(struct drgn_thread *thread, char **ret)
+{
+#define FORMAT "/proc/%" PRIu32 "/comm"
+	char path[sizeof(FORMAT)
+		- sizeof("%" PRIu32)
+		+ max_decimal_length(uint32_t)
+		+ 1];
+	snprintf(path, sizeof(path), FORMAT, thread->tid);
+#undef FORMAT
+	FILE *file = fopen(path, "r");
+	if (!file)
+		return drgn_error_create_os("fopen", errno, path);
+
+	char buf[TASK_COMM_LEN];
+	size_t bytes_read = 0;
+	while (bytes_read < sizeof(buf)) {
+		size_t bytes_read_iter = fread(buf + bytes_read, 1, sizeof(buf) - bytes_read, file);
+		if (bytes_read_iter == 0) {
+			break;
+		}
+		bytes_read += bytes_read_iter;
+	}
+	fclose(file);
```

Comment:
`FILE` buffering does win us anything here, and it has more confusing error handling than using a bare fd. We also already have a retry loop helper for fds. How about something like this:
```suggestion
	_cleanup_close_ int fd = open(path, O_RDONLY);
	if (fd < 0)
		return drgn_error_create_os("open", errno, path);

	char buf[TASK_COMM_LEN];
	ssize_t bytes_read = read_all(fd, buf, sizeof(buf));
	if (bytes_read < 0)
		return drgn_error_create_os("read", errno, path);
```